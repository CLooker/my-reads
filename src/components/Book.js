import React from 'react';

const convertToHttps = str =>
  str.includes('https') ? str : `${str.slice(0, 4)}s${str.slice(4)}`;

const checkForImageLinks = ({ imageLinks }) => (imageLinks ? true : false);

const returnImageAddress = book =>
  checkForImageLinks(book)
    ? convertToHttps(book.imageLinks.thumbnail)
    : encodeURI(
        `https://via.placeholder.com/128x193?text=${book.title.toUpperCase()}`
      );

const returnSelectValue = shelf => (shelf ? true : ' ');

const Book = ({ book, changeBookshelf, displaySearch }) => {
  const { title, id, shelf, authors } = book;
  return (
    <div>
      <div className="book">
        <div className="book-top">
          <div
            className="book-cover"
            style={{
              backgroundImage: `url(${returnImageAddress(book)})`
            }}
          />
          <div className="book-shelf-changer">
            <select
              onChange={changeBookshelf}
              value={returnSelectValue(shelf)}
              data={id}
            >
              <option value="none" disabled>
                Move to...
              </option>
              {displaySearch && <option value=" " disabled />}
              <option value="currentlyReading">Currently Reading</option>
              <option value="wantToRead">Want to Read</option>
              <option value="read">Read</option>
              {shelf && <option value="none">None</option>}
            </select>
          </div>
        </div>
        <div className="book-title">{title}</div>
        <div className="book-authors">
          {authors &&
            authors.map(author => (
              <div key={`${author}${id}`.replace(/\s/g, '')}>{author}</div>
            ))}
        </div>
      </div>
    </div>
  );
};

export default Book;
