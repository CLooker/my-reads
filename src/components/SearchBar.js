import React from 'react';
import { Link } from 'react-router-dom';
import { DebounceInput } from 'react-debounce-input';
import PropTypes from 'prop-types';

const SearchBar = ({ query, handleChange }) => (
  <div className="search-books">
    <div className="search-books-bar">
      <Link className="close-search" to="/my-reads">
        Close
      </Link>
      <div className="search-books-input-wrapper">
        <DebounceInput
          minLength={1}
          debounceTimeout={300}
          type="text"
          placeholder="Search by title or author"
          value={query}
          onChange={handleChange}
        />
      </div>
    </div>
    <div className="search-books-results">
      <ol className="books-grid" />
    </div>
  </div>
);

SearchBar.propTypes = {
  query: PropTypes.string,
  handleChange: PropTypes.func.isRequired
};

export default SearchBar;
