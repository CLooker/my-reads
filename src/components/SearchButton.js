import React from 'react';
import { Link } from 'react-router-dom';

const SearchButton = ({ resetSearch, ...location }) => (
  <div className="open-search">
    <Link onClick={resetSearch} to="my-reads/search">
      Add a book
    </Link>
  </div>
);

export default SearchButton;
