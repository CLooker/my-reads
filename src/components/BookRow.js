import React from 'react';
import Book from './Book';
import PropTypes from 'prop-types';

const BookRow = ({
  myReads,
  shelfTitle,
  shelf,
  changeBookshelf,
  displaySearch
}) => (
  <div className="list-books">
    <div className="list-books-title">
      <h1>{myReads}</h1>
    </div>
    <div className="list-books-content">
      <div className="bookshelf">
        <h2 className="bookshelf-title">{shelfTitle}</h2>
        <div className="bookshelf-books">
          <ol className="books-grid">
            {shelf.map(book => (
              <li key={book.id}>
                <Book
                  book={book}
                  changeBookshelf={changeBookshelf}
                  displaySearch={displaySearch}
                />
              </li>
            ))}
          </ol>
        </div>
      </div>
    </div>
  </div>
);

BookRow.propTypes = {
  myReads: PropTypes.string,
  shelfTitle: PropTypes.string.isRequired,
  shelf: PropTypes.array.isRequired,
  changeBookshelf: PropTypes.func.isRequired,
  displaySearch: PropTypes.bool
};

export default BookRow;
