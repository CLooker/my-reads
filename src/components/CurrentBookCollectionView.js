import React from 'react';
import BookRow from './BookRow';
import SearchButton from './SearchButton';
import PropTypes from 'prop-types';

const CurrentBookCollectionView = ({ returnMyReadsProps, ...routeProps }) => (
  <div>
    <BookRow {...returnMyReadsProps('Currently Reading', routeProps)} />
    <BookRow {...returnMyReadsProps('Want To Read', routeProps)} />
    <BookRow {...returnMyReadsProps('Read', routeProps)} />
    <SearchButton {...routeProps} />
  </div>
);

CurrentBookCollectionView.propTypes = {
  returnMyReadsProps: PropTypes.func.isRequired
};

export default CurrentBookCollectionView;
