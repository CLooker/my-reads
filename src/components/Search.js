import React, { Component } from 'react';
import SearchBar from './SearchBar';
import BookRow from './BookRow';

class Search extends Component {
  render() {
    const {
      closeSearch,
      search,
      handleChange,
      query,
      searchResults,
      changeBookshelf
    } = this.props;
    return (
      <div>
        <SearchBar
          closeSearch={closeSearch}
          search={search}
          handleChange={handleChange}
          query={query}
        />
        <BookRow
          myReads=""
          shelfTitle="Search Results"
          shelf={searchResults}
          changeBookshelf={changeBookshelf}
          displaySearch={true}
        />
      </div>
    );
  }
}

export default Search;
