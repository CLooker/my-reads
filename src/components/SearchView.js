import React, { Component } from 'react';
import SearchBar from './SearchBar';
import BookRow from './BookRow';
import * as BooksAPI from './BooksAPI';
import PropTypes from 'prop-types';

export default class SearchView extends Component {
  static propTypes = {
    shelfKeyValueStore: PropTypes.object.isRequired,
    setStateOnApiReturn: PropTypes.func.isRequired,
    getShelfAndRender: PropTypes.func.isRequired
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    let flag = false;
    const newSearchResults = Object.keys(nextProps.shelfKeyValueStore)
      .map(key => nextProps.shelfKeyValueStore[key])
      .map(book => {
        const target = book.id;
        const match = prevState.searchResults.find(({ id }) => id === target);
        if (match) {
          if (match.shelf !== target.shelf) {
            flag = true;
          }
        }
        return book;
      });

    if (flag === true) {
      return {
        query: prevState.query,
        searchResults: prevState.searchResults.map(book => {
          const target = book.id;
          const newVersion = newSearchResults.find(({ id }) => id === target);
          if (newVersion) {
            return newVersion;
          } else {
            return book;
          }
        })
      };
    } else {
      return null;
    }
  }

  state = {
    query: '',
    searchResults: []
  };

  resetSearch = () => this.setState({ searchResults: [] });

  search = query =>
    query
      ? BooksAPI.search(query).then(this.handleSearchReturn)
      : this.resetSearch();

  handleSearchReturn = res =>
    console.log('search returned this: ', res) ||
    (res !== undefined
      ? this.setState({
          searchResults: res.map(
            bookObj => this.props.shelfKeyValueStore[bookObj.id] || bookObj
          )
        })
      : this.resetSearch());

  changeBookshelf = e =>
    BooksAPI.update(
      { id: e.currentTarget.getAttribute('data') },
      e.target.value.trim()
    ).then(this.props.setStateOnApiReturn);

  handleChange = e =>
    this.setState({ query: e.target.value }, () =>
      this.search(this.state.query)
    );

  returnSearchBarProps = () => ({
    handleChange: this.handleChange,
    query: this.state.query
  });

  returnBookRowProps = () => ({
    shelfTitle: 'Search Results',
    shelf: this.state.searchResults,
    changeBookshelf: this.changeBookshelf,
    displaySearch: true
  });

  render() {
    return (
      <div>
        <SearchBar {...this.returnSearchBarProps()} />
        <BookRow {...this.returnBookRowProps()} />
      </div>
    );
  }
}
