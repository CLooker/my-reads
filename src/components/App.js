import React, { Component } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import * as BooksAPI from './BooksAPI';
import SearchView from './SearchView';
import CurrentBookCollectionView from './CurrentBookCollectionView';
import NoMatch from './NoMatch';
import '../App.css';

export default class App extends Component {
  state = {
    shelf: [],
    shelfKeyValueStore: {},
    currentlyReading: [],
    wantToRead: [],
    read: []
  };

  componentDidMount() {
    this.getShelfAndRender();
  }

  getShelfAndRender() {
    BooksAPI.getAll().then(this.setStateOnApiReturn);
  }

  setStateOnApiReturn = apiReturn =>
    Array.isArray(apiReturn)
      ? this.buildState(apiReturn)
      : this.getShelfAndRender();

  buildState = apiReturn =>
    this.setState({
      shelf: apiReturn,
      shelfKeyValueStore: this.buildKVS(apiReturn),
      currentlyReading: this.buildShelfTypes(apiReturn, 'currentlyReading'),
      wantToRead: this.buildShelfTypes(apiReturn, 'wantToRead'),
      read: this.buildShelfTypes(apiReturn, 'read')
    });

  buildKVS = apiReturn =>
    apiReturn.reduce(
      (keyValueStore, book) => ({
        ...keyValueStore,
        [book.id]: book
      }),
      {}
    );

  buildShelfTypes = (apiReturn, type) =>
    apiReturn.filter(({ shelf }) => shelf === type);

  changeBookshelf = e =>
    console.log('changeBookshelf called') ||
    BooksAPI.update(
      { id: e.currentTarget.getAttribute('data') },
      e.target.value.trim()
    ).then(this.setStateOnApiReturn);

  returnSearchViewProps = routeProps => ({
    shelfKeyValueStore: this.state.shelfKeyValueStore,
    setStateOnApiReturn: this.setStateOnApiReturn,
    getShelfAndRender: this.getShelfAndRender,
    ...routeProps
  });

  returnMyReadsProps = (shelfTitle, routeProps) =>
    shelfTitle === 'Currently Reading'
      ? {
          myReads: 'MyReads',
          shelfTitle,
          shelf: this.state.currentlyReading,
          changeBookshelf: this.changeBookshelf,
          selectedValue: 'currentlyReading',
          ...routeProps
        }
      : shelfTitle === 'Want To Read'
        ? {
            shelfTitle,
            shelf: this.state.wantToRead,
            changeBookshelf: this.changeBookshelf,
            selectedValue: 'wantToRead',
            ...routeProps
          }
        : shelfTitle === 'Read' && {
            shelfTitle,
            shelf: this.state.read,
            changeBookshelf: this.changeBookshelf,
            selectedValue: 'read',
            ...routeProps
          };

  returnCurrentBookCollectionViewProps = routeProps => ({
    returnMyReadsProps: this.returnMyReadsProps,
    ...routeProps
  });

  render() {
    return (
      <Router>
        <div className="app">
          <Switch>
            <Route
              path="/my-reads/search"
              render={routeProps => (
                <SearchView {...this.returnSearchViewProps(routeProps)} />
              )}
            />
            <Route
              exact
              path="/my-reads"
              render={routeProps => (
                <CurrentBookCollectionView
                  {...this.returnCurrentBookCollectionViewProps(routeProps)}
                />
              )}
            />
            <Route component={NoMatch} />
          </Switch>
        </div>
      </Router>
    );
  }
}
