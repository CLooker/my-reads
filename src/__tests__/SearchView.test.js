import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import App from '../components/App';
import SearchView from '../components/SearchView';
import * as BooksAPI from '../components/BooksAPI';

describe('SearchView', () => {
  let searchViewProps, component;

  beforeEach(() => {
    searchViewProps = {
      shelfKeyValueStore: {
        def: {
          id: 'def',
          title: 'fake title 2',
          shelf: 'currentlyReading'
        }
      },
      setStateOnApiReturn: jest.fn(),
      getShelfAndRender: jest.fn()
    };
    component = shallow(<SearchView {...searchViewProps} />);
    fetch.resetMocks();
    jest.restoreAllMocks();
  });

  describe('initialization', () => {
    it('should render properly', () => {
      expect(component).toMatchSnapshot();
    });

    it('should have certain props', () => {
      const mountedComponent = mount(
        <MemoryRouter>
          <SearchView {...searchViewProps} />
        </MemoryRouter>
      );
      const expected = searchViewProps;
      const actual = mountedComponent.find('SearchView').props();
      expect(actual).toEqual(expect.objectContaining(expected));
    });

    it('should have a certain state', () => {
      const expected = {
        query: '',
        searchResults: []
      };

      const actual = component.state();

      expect(actual).toEqual(expected);
    });

    describe('rendering children', () => {
      let el;

      beforeEach(() => {
        el = component.find('SearchBar');
      });
      describe('`SearchBar`', () => {
        it('should render `SearchBar`', () => {
          expect(el).toBeTruthy();
        });

        it('should pass certain props to `SearchBar`', () => {
          const expected = {
            handleChange: el.props().handleChange,
            query: ''
          };
          const actual = el.props();
          expect(actual).toEqual(expect.objectContaining(expected));
        });
      });

      describe('`BookRow`', () => {
        let el;

        beforeEach(() => {
          el = component.find('BookRow');
        });

        it('should render `BookRow`', () => {
          expect(el).toBeTruthy();
        });

        it('should pass certain props to `BookRow`', () => {
          const expected = {
            shelfTitle: 'Search Results',
            shelf: [],
            changeBookshelf: el.props().changeBookshelf,
            displaySearch: true
          };

          const actual = el.props();

          expect(actual).toEqual(expected);
        });
      });
    });
  });

  describe('methods', () => {
    describe('`resetSearch`', () => {
      it('should `setState` properly', () => {
        component.instance().setState({
          searchResults: ['hey', 'ho', "let's go"]
        });

        component.instance().resetSearch();

        const expected = {
          searchResults: []
        };

        const actual = component.state();

        expect(actual).toEqual(expect.objectContaining(expected));
      });
    });

    describe('`search`', () => {
      describe('code paths', () => {
        describe('arg exists', () => {
          describe('`BooksAPI.search` usage', () => {
            const query = 'Android';
            const returnValue = ['hey', 'ho', "let's go"];

            let spy;

            beforeEach(() => {
              BooksAPI.search = jest.fn(async () => await returnValue);
              spy = jest.spyOn(component.instance(), 'handleSearchReturn');
              component.instance().search(query);
            });

            it('should call `BooksAPI.search`', () => {
              expect(BooksAPI.search).toHaveBeenCalled();
            });

            it('should call `BooksAPI.search` with the correct arg', () => {
              const expected = query;
              const actual = BooksAPI.search.mock.calls[0][0];
              expect(actual).toBe(expected);
            });

            describe('`handleSearchReturn` usage', () => {
              it('should call `handleSearchReturn`', () => {
                expect(spy).toHaveBeenCalled();
              });

              it('should call `handleSearchReturn` with the correct arg', () => {
                const expected = returnValue;
                const actual = spy.mock.calls[0][0];
                expect(actual).toEqual(expected);
              });
            });
          });
        });

        describe('arg does not exist', () => {
          it('should call `resetSearch`', () => {
            component.instance().resetSearch = jest.fn();
            const query = '';
            component.instance().search(query);
            expect(component.instance().resetSearch).toHaveBeenCalled();
          });
        });
      });
    });

    describe('`handleSearchReturn`', () => {
      describe('code paths', () => {
        describe('arg is truthy', () => {
          it('should `setState` correctly', () => {
            const res = [
              {
                id: 'abc',
                title: 'fake title 1'
              },
              {
                id: 'def',
                title: 'fake title 2'
              }
            ];

            const expected = {
              searchResults: [
                {
                  id: 'abc',
                  title: 'fake title 1'
                },
                {
                  id: 'def',
                  title: 'fake title 2',
                  shelf: 'currentlyReading'
                }
              ]
            };

            component.instance().handleSearchReturn(res);
            const actual = component.state();
            expect(actual).toEqual(expect.objectContaining(expected));
          });
        });

        describe('arg is falsy', () => {
          it('should call `resetSearch`', () => {
            component.instance().resetSearch = jest.fn();
            const arg = '';
            component.instance().handleSearchReturn(arg);
            expect(component.instance().resetSearch).toHaveBeenCalled();
          });
        });
      });
    });

    describe('`changeBookshelf`', () => {
      const currentTarget = {
        getAttribute: jest.fn(() => '123')
      };

      const target = {
        value: {
          trim: jest.fn(() => 'read')
        }
      };

      beforeEach(() => {
        BooksAPI.update = jest.fn(async () => await []);
        component.instance().changeBookshelf({
          currentTarget,
          target
        });
      });

      describe('fns that should be called', () => {
        describe('`BooksAPI.update` usage', () => {
          it('should `BooksAPI.update`', () => {
            expect(BooksAPI.update).toHaveBeenCalled();
          });

          describe('args', () => {
            describe('first arg', () => {
              it('should be correct', () => {
                const expected = {
                  id: currentTarget.getAttribute()
                };
                const actual = BooksAPI.update.mock.calls[0][0];

                expect(actual).toEqual(expected);
              });
            });

            describe('second arg', () => {
              it('should be correct', () => {
                const expected = target.value.trim();
                const actual = BooksAPI.update.mock.calls[0][1];
                expect(actual).toBe(expected);
              });
            });
          });
        });

        it('should call `setStateOnApiReturn`', () => {
          expect(searchViewProps.setStateOnApiReturn).toHaveBeenCalled();
        });
      });
    });

    describe('`handleChange`', () => {
      let spy;

      const target = {
        value: 'Android'
      };

      const expected = target.value;

      beforeEach(() => {
        spy = jest.spyOn(component.instance(), 'search');
        component.instance().handleChange({ target });
      });

      it('should `setState` properly', () => {
        const actual = component.state().query;
        expect(actual).toBe(expected);
      });

      describe('`search` usage', () => {
        it('should call `search`', () => {
          expect(spy).toHaveBeenCalled();
        });

        it('should call `search` with correct arg', () => {
          const actual = spy.mock.calls[0][0];
          expect(actual).toBe(expected);
        });
      });
    });

    describe('`returnSearchBarProps`', () => {
      it('should return the correct object', () => {
        const expected = {
          handleChange: component.instance().handleChange,
          query: component.state().query
        };

        const actual = component.instance().returnSearchBarProps();

        expect(actual).toEqual(expected);
      });
    });

    describe('`returnBookRowProps`', () => {
      it('should return the correct object', () => {
        const expected = {
          shelfTitle: 'Search Results',
          shelf: component.state().searchResults,
          changeBookshelf: component.instance().changeBookshelf,
          displaySearch: true
        };

        const actual = component.instance().returnBookRowProps();

        expect(actual).toEqual(expected);
      });
    });
  });
});
