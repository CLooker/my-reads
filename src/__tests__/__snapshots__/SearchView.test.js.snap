// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchView initialization should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchView
    getShelfAndRender={[MockFunction]}
    setStateOnApiReturn={[MockFunction]}
    shelfKeyValueStore={
      Object {
        "def": Object {
          "id": "def",
          "shelf": "currentlyReading",
          "title": "fake title 2",
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SearchBar
          handleChange={[Function]}
          query=""
        />,
        <BookRow
          changeBookshelf={[Function]}
          displaySearch={true}
          shelf={Array []}
          shelfTitle="Search Results"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleChange": [Function],
          "query": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "changeBookshelf": [Function],
          "displaySearch": true,
          "shelf": Array [],
          "shelfTitle": "Search Results",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SearchBar
            handleChange={[Function]}
            query=""
          />,
          <BookRow
            changeBookshelf={[Function]}
            displaySearch={true}
            shelf={Array []}
            shelfTitle="Search Results"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleChange": [Function],
            "query": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "changeBookshelf": [Function],
            "displaySearch": true,
            "shelf": Array [],
            "shelfTitle": "Search Results",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
